File: bound/kt/src/test/kotlin/web5/sdk/crypto/keys/InMemoryKeyManagerTest.kt
package web5.sdk.crypto.keys

import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.fail
import web5.sdk.crypto.Ed25519Generator
import web5.sdk.Web5Exception

class InMemoryKeyManagerTest {
  @Nested
  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
  inner class ImportPrivateJwk {

    @Test
    fun test_must_be_private_jwk() {

      val keyManager = InMemoryKeyManager(listOf())
      val privateJwk = Ed25519Generator.generate()
      val publicJwk = privateJwk.copy(d = null)

      val exception = assertThrows<Web5Exception> {
        keyManager.importPrivateJwk(publicJwk)
      }

      assertEquals("parameter error private_jwk must be a private key", exception.message)
      assertEquals("Parameter", exception.variant)
    }

    @Test
    fun test_successfully_imports_and_returns_public_jwk() {

      val keyManager = InMemoryKeyManager(listOf())
      val privateJwk = Ed25519Generator.generate()

      val publicJwk = keyManager.importPrivateJwk(privateJwk)

      assertEquals(publicJwk, privateJwk.copy(d = null))
    }
  }

  @Nested
  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
  inner class GetSigner {
    @Test
    fun test_must_be_public_key() {

      val privateJwk = Ed25519Generator.generate()
      val keyManager = InMemoryKeyManager(listOf(privateJwk))

      val exception = assertThrows<Web5Exception> {
        keyManager.getSigner(privateJwk)
      }

      assertEquals("parameter error public_jwk must be a public key", exception.message)
      assertEquals("Parameter", exception.variant)
    }

    @Test
    fun test_not_found() {

      val keyManager = InMemoryKeyManager(listOf())
      val privateJwk = Ed25519Generator.generate()
      val publicJwk = privateJwk.copy(d = null)

      val exception = assertThrows<Web5Exception> {
        keyManager.getSigner(publicJwk)
      }

      assertEquals("not found error signer not found for public_jwk with thumbprint ${publicJwk.computeThumbprint()}", exception.message)
      assertEquals("NotFound", exception.variant)
    }

    @Test
    fun test_found() {
      
      val privateJwk = Ed25519Generator.generate()
      val publicJwk = privateJwk.copy(d = null)
      val keyManager = InMemoryKeyManager(listOf(privateJwk))

      assertDoesNotThrow {
        keyManager.getSigner(publicJwk)
      }
    }
  }

  @Nested
  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
  inner class ExportPrivateJwks {

    @Test
    fun test_export_empty_list() {

      val keyManager = InMemoryKeyManager(listOf())
      val privateJwks = keyManager.exportPrivateJwks()

      assertTrue(privateJwks.isEmpty())
    }

    @Test
    fun test_export_single_key() {

      val privateJwk = Ed25519Generator.generate()
      val keyManager = InMemoryKeyManager(listOf(privateJwk))
      val privateJwks = keyManager.exportPrivateJwks()

      assertEquals(1, privateJwks.size)
      assertEquals(privateJwk, privateJwks[0])
    }

    @Test
    fun test_export_multiple_keys() {

      val privateJwk1 = Ed25519Generator.generate()
      val privateJwk2 = Ed25519Generator.generate()
      val keyManager = InMemoryKeyManager(listOf(privateJwk1, privateJwk2))
      val privateJwks = keyManager.exportPrivateJwks()

      assertEquals(2, privateJwks.size)
      assertTrue(privateJwks.containsAll(listOf(privateJwk1, privateJwk2)))
    }
  }
}

File: bound/typescript/tests/crypto/key_managers/in_memory_key_manager.test.ts
import { expect } from "chai";
import { InMemoryKeyManager } from "../../../src/crypto/key_managers/in_memory_key_manager";
import { Ed25519Generator } from "../../../src/crypto/dsa/ed25519";
import { Jwk } from "../../../src/crypto/jwk";
import { Web5Error } from "../../../src/errors";

describe("InMemoryKeyManager - importPrivateJwk", () => {
  it("should fail if the JWK is not a private key", async () => {
    const keyManager = new InMemoryKeyManager();
    const privateJwk = Ed25519Generator.generate();
    delete privateJwk.d

    try {
      keyManager.importPrivateJwk(privateJwk);
    } catch (error: any) {
      expect(error instanceof Web5Error).to.equal(true);
      expect(error.variant).to.equal("Parameter");
      expect(error.message).to.equal("parameter error private_jwk must be a private key");
    }
  });

  it("should successfully import and return the public JWK", async () => {
    const keyManager = new InMemoryKeyManager();
    const privateJwk = Ed25519Generator.generate();

    const publicJwk = keyManager.importPrivateJwk(privateJwk);
    expect(publicJwk.d).to.be.undefined;
  });
});

describe("InMemoryKeyManager - getSigner", () => {
  it("should fail if the JWK is not a public key", async () => {
    const keyManager = new InMemoryKeyManager();
    const privateJwk = Ed25519Generator.generate();

    try {
      keyManager.getSigner(privateJwk);
    } catch (error: any) {
      expect(error instanceof Web5Error).to.equal(true);
      expect(error.variant).to.equal("Parameter");
      expect(error.message).to.equal("parameter error public_jwk must be a public key");
    }
  });

  it("should fail if the signer is not found", async () => {
    const keyManager = new InMemoryKeyManager();
    const privateJwk = Ed25519Generator.generate()
    delete privateJwk.d

    try {
      keyManager.getSigner(privateJwk);
    } catch (error: any) {
      expect(error instanceof Web5Error).to.equal(true);
      expect(error.variant).to.equal("NotFound");
      expect(error.message).to.equal(
        `not found error signer not found for public_jwk with thumbprint ${privateJwk.computeThumbprint()}`
      );
    }
  });

  it("should return a signer if the public JWK is found", async () => {
    const keyManager = new InMemoryKeyManager();
    const privateJwk = Ed25519Generator.generate();

    let publicJwk = keyManager.importPrivateJwk(privateJwk);

    const signer = keyManager.getSigner(publicJwk);
    expect(signer).to.not.be.undefined;
  });
});